nfs: Fix inconsistency between glibc and libtirpc

xhernandez opened the pull request:
There's a critical difference between RPC implementation in glibc
and libtirpc.

In libtirpc, svc_run() starts a polling loop that listens from all
registered connections and handles requests. When this is done from
multiple threads, a race can happen where both threads are trying
to process the same connection simultaneously. This causes memory
corruption.

However, in glibc, svc_run() only handles the registered connections
on the current thread, so it's necessary to call svc_run() on all
threads that register services.

This patch fixes that problem by calling svc_run() from all threads
only if libtirpc is not used. Otherwise it's called only once.

Change-Id: I97c3c39a9aad90115e7e23c70884a4ee7769a2dd
Updates: #1009
Signed-off-by: Xavi Hernandez <xhernandez@redhat.com>

xhernandez committed:
nfs: Fix inconsistency between glibc and libtirpc

There's a critical difference between RPC implementation in glibc
and libtirpc.

In libtirpc, svc_run() starts a polling loop that listens from all
registered connections and handles requests. When this is done from
multiple threads, a race can happen where both threads are trying
to process the same connection simultaneously. This causes memory
corruption.

However, in glibc, svc_run() only handles the registered connections
on the current thread, so it's necessary to call svc_run() on all
threads that register services.

This patch fixes that problem by calling svc_run() from all threads
only if libtirpc is not used. Otherwise it's called only once.

Change-Id: I97c3c39a9aad90115e7e23c70884a4ee7769a2dd
Updates: #1009
Signed-off-by: Xavi Hernandez <xhernandez@redhat.com>

mykaul commented on xlators/nfs/server/src/nfs.h:
```diff
@@ -101,6 +101,11 @@ struct nfs_state {
+
+#ifdef HAVE_LIBTIRPC
+    pthread_mutex_t svc_mutex;
+    bool svc_running;
```
Out of curiosity, could we have used an atomic and GF_ATOMIC_BASE_CMP_SWAP() instead?

xhernandez replied:
Yes, you are right. I'll update the patch.

xhernandez committed:
nfs: Use an atomic instead of a mutex

Instead of using a mutex to synchronize the call to svc_run(), we
use an atomic operation to update the flag.

Change-Id: I401481d1d0308a557271f99a32527a2ab3c4542f
Signed-off-by: Xavi Hernandez <xhernandez@redhat.com>

mykaul commented on xlators/nfs/server/src/nfs.h:
```diff
@@ -101,6 +101,10 @@ struct nfs_state {
     uint32_t event_threads;
+
+#ifdef HAVE_LIBTIRPC
+    bool svc_running;
```
nit: it's not like the structure is anyway aligned, but this is not helping ;-)

xhernandez replied:
Why not ? I've already put it the last field of the structure. A `bool` is a single byte type. All other types (including `gf_boolean_t`) are at least 4 bytes. I don't see how I can improve the layout of the structure without reordering all fields.

mykaul replied:
I thought attaching it to another boolean would have helped a bit.

xhernandez replied:
I've just realized that `gf_boolean_t` is now an alias of `bool`. Previously it was an `enum`. In this case, yes, I could reorder the fields a bit. If I do that, wouldn't it be better to reorder all fields instead of only bools ?

mykaul replied:
I would say yes, of course. If we are already touching a structure, let's do the janitor work and fix alignment. Others may disagree. But it's a nit, really, so no strong feels either way.

xhernandez committed:
nfs: Align fields of struct nfs_state

Change-Id: Ib0da4911533ffd9c08636457b664eaa45172e975
Signed-off-by: Xavi Hernandez <xhernandez@redhat.com>

mohit84 commented:
/run regression

xhernandez commented:
/run regression
